add reagent example

gadfly361 opened the pull request:
Submitting example for [reagent](https://github.com/reagent-project/reagent) in relation to #985.  

Reagent provides a minimalistic interface between ClojureScript and React. It allows you to define efficient React components using nothing but plain ClojureScript functions and data, that describe your UI using a Hiccup-like synta

  TodoMVC - reagent
    When page is initially opened
      ✓ should focus on the todo input field (61ms)
    No Todos
      ✓ should hide #main and #footer (267ms)
    New Todo
      ✓ should allow me to add todo items (872ms)
      ✓ should clear text input field when an item is added (376ms)
      ✓ should append new items to the bottom of the list (1265ms)
      ✓ should trim text input (400ms)
      ✓ should show #main and #footer when items added (537ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (1184ms)
      ✓ should allow me to clear the completion state of all items (1570ms)
      ✓ complete all checkbox should update state when items are completed / cleared (2060ms)
    Item
      ✓ should allow me to mark items as complete (810ms)
      ✓ should allow me to un-mark items as complete (1283ms)
      ✓ should allow me to edit an item (2748ms)
      ✓ should show the remove button on hover 
    Editing
      ✓ should hide other controls when editing (1761ms)
      ✓ should save edits on enter (2992ms)
      ✓ should save edits on blur (3606ms)
      ✓ should trim entered text (2522ms)
      ✓ should remove the item if an empty text string was entered (2385ms)
      ✓ should cancel edits on escape (2225ms)
    Counter
      ✓ should display the current number of todo items (925ms)
    Clear completed button
      ✓ should display the number of completed items (2149ms)
      ✓ should remove completed items when clicked (927ms)
      ✓ should be hidden when there are no items that are completed (1034ms)
    Routing
      ✓ should allow me to display active items (1984ms)
      ✓ should respect the back button (2295ms)
      ✓ should allow me to display completed items (1390ms)
      ✓ should allow me to display all items (1993ms)
      ✓ should highlight the currently applied filter (2317ms)


  29 passing (1m)

samccone commented on examples/reagent/src/cljs/todomvc/actions.cljs:
```diff
@@ -0,0 +1,39 @@
+      (helpers/todos-all todos))))
+
+(defn clear-completed-todos [todos]
+    (doseq [todo (helpers/todos-completed todos)]
```
1 too many indents

gadfly361 replied:
yup, also looks like line 28 needs to be indented.

samccone commented on examples/reagent/src/cljs/todomvc/components/todo_edit.cljs:
```diff
@@ -0,0 +1,30 @@
+(defn component-render [{:keys [id title completed]} editing]
+  (let [default title
+        edit-title (reagent/atom default)]
+    (fn []
```
might be nice to use 

`#()` here 

thoughts?

gadfly361 replied:
I think the reagent community has defaulted to using (fn [ ] element) instead of #( ) in cases like this.

samccone replied:
kk I defer to community best practices 

was approaching this only from a clojure perspective :)

wkf replied:
Using `#()` wouldn't be possible anyway, since they can't be nested. :)

samccone commented on examples/reagent/src/cljs/todomvc/components/todo_item.cljs:
```diff
@@ -0,0 +1,23 @@
+            [todomvc.components.todo-edit :as todo-edit]))
+
+(defn todo-item-class [completed editing]
+  (str (if completed "completed ")
```
ha this is funny :tongue:

samccone replied:
but it is good, just clever

wkf replied:
Why `if` over `when` ?

gadfly361 replied:
:+1: for `when`

passy commented:
:+1: for including. I've looked at it before and it captures the "React spirit" pretty well in cljs.

addyosmani commented:
:+1: for this landing once we're happy with the code review.

samccone commented:
@gadfly361 Looks like you forgot to modify https://github.com/tastejs/todomvc/blob/master/learn.json

Update that with the reagent info and then squash this PR into a single commit and we should be good to go

:+1: thanks

arthurvr commented on examples/reagent/.gitignore:
```diff
@@ -0,0 +1,9 @@
+
+node_modules/todomvc-common/*
+!node_modules/todomvc-common/base.css
+!node_modules/todomvc-common/base.js
```
Please add a final newline

arthurvr commented on examples/reagent/index.html:
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
```
lowercase the doctype: `<!doctype html>`

arthurvr commented:
Please also add the app to the home page (`index.html`)

arthurvr commented on examples/reagent/index.html:
```diff
@@ -0,0 +1,18 @@
+    <script>
+      todomvc.core.run();
+    </script>
+  </body>
```
The indentation in this file is wrong. We use tabs.

arthurvr commented on examples/reagent/index.html:
```diff
@@ -0,0 +1,18 @@
+  <body>
+    <div id="app"> Loading... </div>
+    <script src="node_modules/todomvc-common/base.js"></script>
+    <script src="js/app.js"></script>
```
That file will need to be included in the repo.

arthurvr replied:
Without the app isn't able to run on the live site.

arthurvr commented on examples/reagent/index.html:
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
+<html lang="en">
```
Please add a  `data-framework` similar to the other apps.

arthurvr commented on examples/reagent/index.html:
```diff
@@ -0,0 +1,18 @@
+    <link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+  </head>
+  <body>
+    <div id="app"> Loading... </div>
```
Are the spaces around the `Loading...` there for a particular reason?

gadfly361 replied:
Nope

arthurvr commented on examples/reagent/package.json:
```diff
@@ -0,0 +1,7 @@
+    "dependencies": {
+	"todomvc-app-css": "^1.0.0",
+	"todomvc-common": "^1.0.1"
+    }
```
Wrong indentation. `package.json` files all use 2 spaces: https://github.com/tastejs/todomvc/blob/master/.editorconfig#L10

arthurvr commented on examples/reagent/project.clj:
```diff
@@ -0,0 +1,21 @@
+(defproject todomvc "0.1.0-SNAPSHOT"
+  :dependencies [[org.clojure/clojure "1.6.0"]
```
Is the spaces indentation a requirement? All our code should use tabs if that's possible.

arthurvr commented on examples/reagent/README.md:
```diff
@@ -0,0 +1,45 @@
+
+> Reagent provides a minimalistic interface between ClojureScript and React. It allows you to define efficient React components using nothing but plain ClojureScript functions and data, that describe your UI using a Hiccup-like syntax.
+
+> _[reagent - https://github.com/reagent-project/reagent](https://github.com/reagent-project/reagent)_
```
`Reagent`

arthurvr commented on examples/reagent/README.md:
```diff
@@ -0,0 +1,45 @@
+
+# Learning Reagent
+
+The [reagent website](https://github.com/reagent-project/reagent) is a great resource for getting started.
```
`Reagent`

arthurvr replied:
> [reagent website](https://github.com/reagent-project/reagent)

That's the github repo, the website is http://reagent-project.github.io/.

arthurvr commented on examples/reagent/README.md:
```diff
@@ -0,0 +1,45 @@
+
+# Credit
+
+[Gadfly361](https://twitter.com/gadfly361), [Dmitri Sotnikov](https://twitter.com/yogthos), and [Dan Holmsand](https://twitter.com/holmsand)
```
> Gadfly361

Inconsistency: for other people you include their full name.

arthurvr commented on examples/reagent/index.html:
```diff
@@ -0,0 +1,18 @@
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>Reagent • TodoMVC</title>
```
Please move the title right under the `<meta charset="utf-8">` for consistency with the other apps.

arthurvr commented:
Left some comments, thanks for working on this! Will manually test again when you included the required `app.js` file.

arthurvr commented on examples/reagent/README.md:
```diff
@@ -0,0 +1,45 @@
+
+> _[reagent - https://github.com/reagent-project/reagent](https://github.com/reagent-project/reagent)_
+
+# Learning Reagent
```
Make this all sub-headings.

gadfly361 commented:
@arthurvr Addressed comments and added the `app.js` file.  

@samccone updated `learn.json`. Will squash into single commit once I get thumbs up from @arthurvr.

arthurvr commented on examples/reagent/src/cljs/todomvc/actions.cljs:
```diff
@@ -0,0 +1,39 @@
+(ns todomvc.actions
+  (:require [todomvc.session :as session]
```
This one is still 2 spaces?

arthurvr replied:
Some other files as well. Package.json is the only one that shouldn't use tabs.

gadfly361 replied:
Ah, misunderstood, thought you were just referring to `project.clj`. I will go back and tab indent all of the files in `src/`

arthurvr replied:
Thanks! If there's no special reason to keep them on spaces we'd prefer that.

gadfly361 replied:
Please ignore my last comment.  After some more thought, to answer your original question:

*Is the spaces indentation a requirement? All our code should use tabs if that's possible.*

I believe the answer is yes, spaces indentation is a requirement. The following recommend against tabs for clojure:
* https://github.com/bbatsov/clojure-style-guide
* https://clojuredocs.org/examples-styleguide
* http://en.wikibooks.org/wiki/Learning_Clojure/Coding_Conventions

As a result, I will untab `project.clj`.  My apologies for not originally thinking this through.

arthurvr replied:
Yeah. That's why I asked it. Though we normally prefer tabs, it might be a community best-practice, then I guess we're fine with it.

arthurvr commented on learn.json:
Super nitpick but link to `https`, saves the user a redirect :)

arthurvr commented:
I'm seeing the following test failures, using latest chrome:

```
~/code/tastejs/todomvc/browser-tests pr/1250 6s
❯ npm run test -- --framework=reagent

  TodoMVC - reagent
    New Todo
      1) should append new items to the bottom of the list
    Item
      2) should allow me to mark items as complete
      3) should allow me to un-mark items as complete

  26 passing (49s)
  3 failing
```

Would you mind double checking on chrome?

gadfly361 commented:
@arthurvr Looking into the test failures

passy commented:
Thanks for the work you two. Super excited about landing this!

gadfly361 commented:
After updating when the todos are filtered, I ran the tests in Chrome Version 41.0.2272.118 (64-bit) on LinuxMint17.

 TodoMVC - reagent
    When page is initially opened
      ✓ should focus on the todo input field (53ms)
    No Todos
      ✓ should hide #main and #footer (118ms)
    New Todo
      ✓ should allow me to add todo items (561ms)
      ✓ should clear text input field when an item is added (247ms)
      ✓ should append new items to the bottom of the list (752ms)
      ✓ should trim text input (300ms)
      ✓ should show #main and #footer when items added (274ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (658ms)
      ✓ should allow me to clear the completion state of all items (721ms)
      ✓ complete all checkbox should update state when items are completed / cleared (818ms)
    Item
      ✓ should allow me to mark items as complete (602ms)
      ✓ should allow me to un-mark items as complete (562ms)
      ✓ should allow me to edit an item (1084ms)
      ✓ should show the remove button on hover 
    Editing
      ✓ should hide other controls when editing (588ms)
      ✓ should save edits on enter (1057ms)
      ✓ should save edits on blur (1122ms)
      ✓ should trim entered text (1058ms)
      ✓ should remove the item if an empty text string was entered (970ms)
      ✓ should cancel edits on escape (1042ms)
    Counter
      ✓ should display the current number of todo items (453ms)
    Clear completed button
      ✓ should display the number of completed items (791ms)
      ✓ should remove completed items when clicked (816ms)
      ✓ should be hidden when there are no items that are completed (673ms)
    Routing
      ✓ should allow me to display active items (794ms)
      ✓ should respect the back button (982ms)
      ✓ should allow me to display completed items (785ms)
      ✓ should allow me to display all items (893ms)
      ✓ should highlight the currently applied filter (738ms)


  29 passing (50s)

arthurvr commented:
Hmmm. I still can't get them passing (8 failing ones). Can somebody else try?

samccone commented:
@arthurvr all passing on my system


```bash
  TodoMVC - reagent
    When page is initially opened
      ✓ should focus on the todo input field
    No Todos
      ✓ should hide #main and #footer (84ms)
    New Todo
      ✓ should allow me to add todo items (536ms)
      ✓ should clear text input field when an item is added (255ms)
      ✓ should append new items to the bottom of the list (742ms)
      ✓ should trim text input (316ms)
      ✓ should show #main and #footer when items added (281ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (765ms)
      ✓ should allow me to clear the completion state of all items (871ms)
      ✓ complete all checkbox should update state when items are completed / cleared (768ms)
    Item
      ✓ should allow me to mark items as complete (642ms)
      ✓ should allow me to un-mark items as complete (643ms)
      ✓ should allow me to edit an item (1106ms)
      ✓ should show the remove button on hover
    Editing
      ✓ should hide other controls when editing (645ms)
      ✓ should save edits on enter (1005ms)
      ✓ should save edits on blur (1072ms)
      ✓ should trim entered text (995ms)
      ✓ should remove the item if an empty text string was entered (858ms)
      ✓ should cancel edits on escape (914ms)
    Counter
      ✓ should display the current number of todo items (425ms)
    Clear completed button
      ✓ should display the number of completed items (737ms)
      ✓ should remove completed items when clicked (719ms)
      ✓ should be hidden when there are no items that are completed (707ms)
    Routing
      ✓ should allow me to display active items (767ms)
      ✓ should respect the back button (1487ms)
      ✓ should allow me to display completed items (832ms)
      ✓ should allow me to display all items (1035ms)
      ✓ should highlight the currently applied filter (1294ms)


  29 passing (1m)
```

arthurvr commented:
@sindresorhus You've got anything else or is this good to go?

gadfly361 commented:
Just squashed commits and should be ready to go. Let me know if you need anything else changed and I'll be happy to update.  Really excited to be able to contribute to todomvc! :)

arthurvr commented:
> Just squashed commits and should be ready to go. Let me know if you need anything else changed and I'll be happy to update. Really excited to be able to contribute to todomvc! :)

Awesome, thanks! We're just waiting to give other people a chance to take a look.

samccone commented:
ping @tastejs/todomvc any one else want to take a look?

passy commented:
Yup, will take a look now.

passy commented on index.html:
```diff
@@ -164,6 +164,9 @@
 									<a href="examples/serenadejs/" data-source="https://github.com/elabs/serenade.js" data-content="Serenade.js is yet another MVC client side JavaScript framework. Why do we indulge in recreating the wheel? We believe that Serenade.js more closely follows the ideas of classical MVC than competing frameworks.">Serenade.js</a>
 								</li>
+								<li class="routing">
+									<a href="examples/reagent/" data-source="https://github.com/reagent-project/reagent" data-content="Reagent provides a minimalistic interface between ClojureScript and React.">Reagent</a>
```
I think linking to `https://reagent-project.github.io/` would be better.

passy commented on learn.json:
```diff
@@ -1813,6 +1813,43 @@
+	"reagent": {
+		"name": "Reagent",
+		"description": "Reagent provides a minimalistic interface between ClojureScript and React. It allows you to define efficient React components using nothing but plain ClojureScript functions and data, that describe your UI using a Hiccup-like syntax.",
+		"homepage": "http://reagent-project.github.io/",
```
This must be without the leading `http://` for some legacy reasons.

passy commented:
Maybe I'm missing something, but is it possible that the persistence layer is missing? When I refresh the page, the items are gone and I couldn't find anything in the source to handle persistent storage.

gadfly361 commented:
@passy I updated the links as you suggested.  Also, you aren't missing anything, I didn't include persistence because the tests were commented out, see [here](https://github.com/tastejs/todomvc/blob/master/browser-tests/test.js#L290).  Should those tests not be commented out?

arthurvr commented:
@gadfly361 The [app spec](https://github.com/tastejs/todomvc/blob/master/app-spec.md#persistence) still has it, so every app should still have it. The test were commented out because there were problems with the tests (I think?)

samccone commented:
yeah it looks like it was commented out just because they were flakey, worth investigating again :)

gadfly361 commented:
Will add persistence. :)

arthurvr commented:
Thanks!

gadfly361 commented:
Added local storage

  TodoMVC - reagent
    When page is initially opened
      ✓ should focus on the todo input field (51ms)
    No Todos
      ✓ should hide #main and #footer (128ms)
    New Todo
      ✓ should allow me to add todo items (490ms)
      ✓ should clear text input field when an item is added (298ms)
      ✓ should append new items to the bottom of the list (594ms)
      ✓ should trim text input (234ms)
      ✓ should show #main and #footer when items added (250ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (619ms)
      ✓ should allow me to clear the completion state of all items (703ms)
      ✓ complete all checkbox should update state when items are completed / cleared (808ms)
    Item
      ✓ should allow me to mark items as complete (518ms)
      ✓ should allow me to un-mark items as complete (540ms)
      ✓ should allow me to edit an item (967ms)
      ✓ should show the remove button on hover 
    Editing
      ✓ should hide other controls when editing (588ms)
      ✓ should save edits on enter (913ms)
      ✓ should save edits on blur (1000ms)
      ✓ should trim entered text (1030ms)
      ✓ should remove the item if an empty text string was entered (894ms)
      ✓ should cancel edits on escape (907ms)
    Counter
      ✓ should display the current number of todo items (425ms)
    Clear completed button
      ✓ should display the number of completed items (684ms)
      ✓ should remove completed items when clicked (665ms)
      ✓ should be hidden when there are no items that are completed (703ms)
    Routing
      ✓ should allow me to display active items (710ms)
      ✓ should respect the back button (962ms)
      ✓ should allow me to display completed items (669ms)
      ✓ should allow me to display all items (773ms)
      ✓ should highlight the currently applied filter (762ms)


  29 passing (48s)

arthurvr commented on examples/reagent/README.md:
```diff
@@ -0,0 +1,45 @@
+# Reagent TodoMVC Example
```
Could you rename this file to `readme.md` (lowercase)? To be consistent with other apps.

gadfly361 replied:
Certainly! Just updated it.

arthurvr commented:
Punting the review to somebody that knows a little more about ClojureScript but I tested and the localstorage works fine.

gadfly361 commented:
Thanks @arthurvr et al for all the help so far! Looking forward to resolving any outstanding issues :)

passy commented on examples/reagent/src/cljs/todomvc/helpers.cljs:
```diff
@@ -0,0 +1,42 @@
+    true))
+
+(defn create-todo-id []
+  (swap! session/todos-counter inc))
```
Based on the spec(*), we use a UUID instead of an incrementing counter for the identifier.

(*) I actually have to check if we require it there, but it's the de facto way to implement it.

gadfly361 replied:
I don't believe that is a requirement of the spec

passy commented on examples/reagent/src/cljs/todomvc/session.cljs:
```diff
@@ -0,0 +1,12 @@
+(ns todomvc.session
+  (:require [reagent.core :as reagent]
+            [alandipert.storage-atom :refer [local-storage]]))
```
It's a little unfortunate that the library doesn't support namespacing at all. We usually require apps to have a special key in localStorage. It seems like you'd have to implement your own StorageBackend with this. The problem is, that if we don't do it this time, the next example that tries to make use of the library here will break. Not quite sure what the best course of action would be.

gadfly361 replied:
I haven't used this library before so I could be wrong, but I think the key for the todos will become "todos-reagent" from line 5. I should probably rename the key for the counter from "counter' to "counter-reagent" or something though.

gadfly361 replied:
Renamed key for counter to have reagent in the name

samccone commented:
Hey @gadfly361 one more tiny thing we have dropped the completed count from the application spec so would you mind removing it from this example and then we should be good to go

gadfly361 commented:
@samccone Will do :)

gadfly361 commented:
Completed count has been removed. 

  TodoMVC - reagent
    When page is initially opened
      ✓ should focus on the todo input field (59ms)
    No Todos
      ✓ should hide #main and #footer (108ms)
    New Todo
      ✓ should allow me to add todo items (1043ms)
      ✓ should clear text input field when an item is added (609ms)
      ✓ should append new items to the bottom of the list (960ms)
      ✓ should trim text input (899ms)
      ✓ should show #main and #footer when items added (576ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (1771ms)
      ✓ should allow me to clear the completion state of all items (1655ms)
      ✓ complete all checkbox should update state when items are completed / cleared (1366ms)
    Item
      ✓ should allow me to mark items as complete (1031ms)
      ✓ should allow me to un-mark items as complete (1401ms)
      ✓ should allow me to edit an item (3146ms)
      ✓ should show the remove button on hover 
    Editing
      ✓ should hide other controls when editing (1238ms)
      ✓ should save edits on enter (2413ms)
      ✓ should save edits on blur (3009ms)
      ✓ should trim entered text (1638ms)
      ✓ should remove the item if an empty text string was entered (2171ms)
      ✓ should cancel edits on escape (2755ms)
    Counter
      ✓ should display the current number of todo items (685ms)
    Clear completed button
      ✓ should remove completed items when clicked (1621ms)
      ✓ should be hidden when there are no items that are completed (1382ms)
    Routing
      ✓ should allow me to display active items (2057ms)
      ✓ should respect the back button (2436ms)
      ✓ should allow me to display completed items (1388ms)
      ✓ should allow me to display all items (1808ms)
      ✓ should highlight the currently applied filter (2420ms)


  28 passing (1m)

samccone commented:
in it goes

thanks again @gadfly361 !!!

gadfly361 commented:
Thank you all for your feedback and for adding reagent to TodoMVC!! Very excited :)

passy commented:
![](http://media.giphy.com/media/UkhHIZ37IDRGo/giphy.gif)
